### PR TITLE
Nest stakeholders in contracts table as JSON arrays

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
changelog_begin
changelog_end

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by stefanobaghino-da** [comment]
There seems to be an error with a `SELECT` clause that raises some 1011 "Invalid column index" error, but I cannot spot it.

File: ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala [hunk 1]
```diff
@@ -645,41 +621,17 @@ private object OracleQueries extends Queries {
...
+        INTO contract (contract_id, tpid, key, payload, signatories, observers, agreement_text)
+        VALUES (?, ?, ?, ?, ?, ?, ?)
       """,
       logHandler0 = log,
     ).updateMany(
-      dbcs
-        .map { c =>
-          (c.contractId, c.templateId, c.key, c.payload, c.agreementText)
-        }
+      dbcs.map(c => c.copy(signatories = c.signatories.toJson, observers = c.observers.toJson))
```
Comment (S11001001) [inline ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala:633]:
```suggestion
      dbcs.map(_.mapKeyPayloadParties(identity, identity, _.toJson))
```

File: ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala [hunk 2]
```diff
@@ -690,6 +642,26 @@ private object OracleQueries extends Queries {
...
+        case JsArray(values) => values.traverse(extractString)
+        case _ => None
+      }
+    }
+    def assertStringArray(column: String, value: JsValue): Vector[String] =
+      extractStringArray(value).getOrElse(
+        sys.error(
+          s"Column '$column' should contain an array of strings, ${value.compactPrint} found instead"
+        )
+      )
```
Comment (S11001001) [inline ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala:663]:
```suggestion
      value.convertTo[Vector[String]]
```

File: ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala [hunk 3]
```diff
@@ -705,40 +677,24 @@ private object OracleQueries extends Queries {
...
-      // see https://github.com/digital-asset/daml/issues/9388#issuecomment-820538688
-      // for a demonstration
-      val q = sql"SELECT $outerSelectList FROM ($dupQ) WHERE rownumber = 1"
+      // Oracle SQL/JSON path expression syntax: https://docs.oracle.com/en/database/oracle/oracle-database/19/adjsn/json-path-expressions.html#GUID-AEBAD813-99AB-418A-93AB-F96BC1658618
+      val quotedParties = parties.toVector.map(p => s""""$p"""").mkString(", ")
+      val q =
+        sql"""SELECT c.contract_id contract_id, $tpid template_id, key, payload, signatories, observers, agreement_text
+                FROM contract c
+                WHERE (json_exists(signatories, '$$[*]?(@ in ($quotedParties))')
+                       OR json_exists(observers, '$$[*]?(@ in ($quotedParties))')
```
Comment (S11001001) [inline ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala:686]: The key question for this PR: can we index the database such that these conditions are not applied linearly?

Also I strongly suggest two modifications to your path construction:

1. Put them together with `Cord` and pass the whole path through `oracleShortPathEscape`.
2. If possible, pass `quotedParties` using the `PASSING ? AS X` binding syntax used elsewhere in this file.

File: ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala [hunk 4]
```diff
@@ -690,6 +642,26 @@ private object OracleQueries extends Queries {
...
+        case _ => None
+      }
+    }
+    def extractStringArray(value: JsValue): Option[Vector[String]] = {
+      import scalaz.syntax.traverse._
+      value match {
+        case JsArray(values) => values.traverse(extractString)
+        case _ => None
+      }
+    }
```
Comment (S11001001) [inline ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala:657]:
```suggestion
```

File: ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala (hunk 1 above)
Comment (stefanobaghino-da) [reply ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala]: Addressed by https://github.com/digital-asset/daml/pull/9484/commits/ab4d51eff82d1d06c9708981bb65be30897dbf1b

File: ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala (hunk 2 above)
Comment (stefanobaghino-da) [reply ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala]: Addressed by https://github.com/digital-asset/daml/pull/9484/commits/35bb16a72b225e28d564805cdc087f29e4603f10

File: ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala (hunk 4 above)
Comment (stefanobaghino-da) [reply ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala]: Addressed by https://github.com/digital-asset/daml/pull/9484/commits/35bb16a72b225e28d564805cdc087f29e4603f10

File: ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala (hunk 3 above)
Comment (stefanobaghino-da) [reply ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala]: Regarding the changes to the code: I've addressed the first point in https://github.com/digital-asset/daml/pull/9484/commits/b285faee00cb844cc44e9d26f689f85747106e22, as for the second I've done a few experiments but nothing worked and I always ended up with something which didn't work. I believe I got close to something working but I'm not sure of how to interpolate the parties vector as a list of values (e.g. `("Alice", "Bob")`).

**Comment by stefanobaghino-da** [comment]
> There seems to be an error with a `SELECT` clause that raises some 1011 "Invalid column index" error, but I cannot spot it.

Ultimately I believe the issue was a very well hidden missing parentheses that unfortunately caused this very misleading error to appear. Now all tests pass locally.

File: ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala (hunk 3 above)
Comment (stefanobaghino-da) [reply ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala]: As for the main point, I'm doing a bit of experimentation, but I believe the [JSON Search Index](https://docs.oracle.com/en/database/oracle/oracle-database/19/adjsn/indexes-for-json-data.html#GUID-D7A604E1-F617-4C92-A9D9-00B6AB78EFD8) you brought up should do the job.

**Comment by S11001001** [comment]
b285fae with #9492 (no extra indices):

```
---- SyncQueryRequest ----------------------------------------------------------
> Number of requests                                   500 (OK=500    KO=-     )
> Min. response time                                    22 (OK=22     KO=-     )
> Max. response time                                  1588 (OK=1588   KO=-     )
> Mean response time                                    34 (OK=34     KO=-     )
> Std. deviation                                        70 (OK=70     KO=-     )
> response time 90th percentile                         38 (OK=38     KO=-     )
> response time 95th percentile                         43 (OK=43     KO=-     )
> response time 99th percentile                         59 (OK=59     KO=-     )
> response time 99.9th percentile                     1588 (OK=1588   KO=-     )
> Mean requests/second                              29.119 (OK=29.119 KO=-     )
```

File: ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala [hunk 5]
```diff
@@ -705,40 +658,27 @@ private object OracleQueries extends Queries {
...
-                          ON (c.contract_id = od.contract_id)
-              WHERE (${Fragments.in(fr"sm.party", parties)}
-                     OR ${Fragments.in(fr"om.party", parties)})
-                    AND $queriesCondition"""
-      // see https://github.com/digital-asset/daml/issues/9388#issuecomment-820538688
-      // for a demonstration
-      val q = sql"SELECT $outerSelectList FROM ($dupQ) WHERE rownumber = 1"
+      val quotedParties = parties.toVector.map(p => s""""$p"""").mkString(", ")
+      val partiesQuery = oracleShortPathEscape(
+        '$' -: Cord.stringToCord("[*]?(@ in (") :+ quotedParties :+ "))"
```
Comment (S11001001) [inline ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala:663]:
```suggestion
        '$' -: ("[*]?(@ in (": Cord) :+ quotedParties :+ "))"
```

**Review by S11001001 — APPROVED** [review]
Approved modulo included suggestion.

File: ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala (hunk 3 above)
Comment (S11001001) [reply ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala]: > I believe I got close to something working but I'm not sure of how to interpolate the parties vector as a list of values

The trouble as it stands is that if the sum of lengths of party IDs in the request is >4000 bytes or so, you simply won't be able to query.

**Comment by stefanobaghino-da** [comment]
> [b285fae](https://github.com/digital-asset/daml/commit/b285faee00cb844cc44e9d26f689f85747106e22) with #9492 (no extra indices):
> 
> ```
> ---- SyncQueryRequest ----------------------------------------------------------
> > Number of requests                                   500 (OK=500    KO=-     )
> > Min. response time                                    22 (OK=22     KO=-     )
> > Max. response time                                  1588 (OK=1588   KO=-     )
> > Mean response time                                    34 (OK=34     KO=-     )
> > Std. deviation                                        70 (OK=70     KO=-     )
> > response time 90th percentile                         38 (OK=38     KO=-     )
> > response time 95th percentile                         43 (OK=43     KO=-     )
> > response time 99th percentile                         59 (OK=59     KO=-     )
> > response time 99.9th percentile                     1588 (OK=1588   KO=-     )
> > Mean requests/second                              29.119 (OK=29.119 KO=-     )
> ```

So, if both schemas don't have indices, flattening the stakeholders in the contracts table improves the performance, right?

File: ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala (hunk 3 above)
Comment (stefanobaghino-da) [reply ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala]: A JSON Search Index gets the job done, I tested it on a table with 10K items, each having from 1 to 10 parties in a JSON CLOB column. The query sped up significantly and the execution plan reflected the usage of the JSON Search Index I created.

**Comment by cocreature** [comment]
Nice work! Simpler and faster

**Pull request merged**